pkg/controller/build/build_controller.go

```diff
@@ -239,67 +242,65 @@ func (r *ReconcileBuild) Reconcile(request reconcile.Request) (reconcile.Result,
 
 	// Populate the status struct with default values
 	b.Status.Registered = corev1.ConditionFalse
-	b.Status.Reason = succeedStatus
+	b.Status.Reason = build.SucceedStatus
+	validationFailed = false
 
 	// Validate if remote repository exists
 	if b.Spec.Source.SecretRef == nil {
 		if err := r.validateSourceURL(ctx, b, b.Spec.Source.URL); err != nil {
-			b.Status.Reason = err.Error()
-			if updateErr := r.client.Status().Update(ctx, b); updateErr != nil {
-				return reconcile.Result{}, updateErr
-			}
-			return reconcile.Result{}, nil
+			validationFailed = true
+			MarkBuildStatus(b, build.RemoteRepositoryUnreachable, err.Error())
+		}
+
+		if validationFailed {
+			return r.UpdateBuildStatusAndRetreat(ctx, b)
 		}
 	}
 
 	// Validate if the referenced secrets exist in the namespace
-	var secretNames []string
+	secretRefMap := map[string]build.BuildReason{}
 	if b.Spec.Output.SecretRef != nil && b.Spec.Output.SecretRef.Name != "" {
-		secretNames = append(secretNames, b.Spec.Output.SecretRef.Name)
+		secretRefMap[b.Spec.Output.SecretRef.Name] = build.SpecOutputSecretRefNotFound
 	}
 	if b.Spec.Source.SecretRef != nil && b.Spec.Source.SecretRef.Name != "" {
-		secretNames = append(secretNames, b.Spec.Source.SecretRef.Name)
+		secretRefMap[b.Spec.Source.SecretRef.Name] = build.SpecSourceSecretRefNotFound
 	}
 	if b.Spec.BuilderImage != nil && b.Spec.BuilderImage.SecretRef != nil && b.Spec.BuilderImage.SecretRef.Name != "" {
-		secretNames = append(secretNames, b.Spec.BuilderImage.SecretRef.Name)
+		secretRefMap[b.Spec.BuilderImage.SecretRef.Name] = build.SpecRuntimeSecretRefNotFound
 	}
 
-	if len(secretNames) > 0 {
-		if err := r.validateSecrets(ctx, secretNames, b.Namespace); err != nil {
-			b.Status.Reason = err.Error()
-			if updateErr := r.client.Status().Update(ctx, b); updateErr != nil {
-				// return an error in case of transient failure, and expect the next
-				// reconciliation to be able to update the Status of the object
-				return reconcile.Result{}, fmt.Errorf("errors: %v %v", err, updateErr)
-			}
-			// The Secret Resource watcher will Reconcile again once the missing
-			// secret is created, therefore no need to return an error and enter on an infinite
-			// reconciliation
-			return reconcile.Result{}, nil
+	// Validate if the referenced secrets exist
+	if len(secretRefMap) > 0 {
+		if err := r.validateSecrets(ctx, secretRefMap, b); err != nil {
+			return reconcile.Result{}, err
+		}
+
+		if validationFailed {
```

Comment:
`validationFailed` here is kind of a global variable. Took me some time to find out how this works. Can `validateSecrets` and `validateStrategyRef` return this similar to `validateRuntimeFailed` ?

Reply:
@SaschaSchwarze0 
I think they can't. Because, the value of `validationFailed` is `false` by default. If `validateSourceURL` failed, then `validationFailed` is changed to be `true`, build status is updated and Build controller stops reconciling. 
The process of `validateSecrets `, `validateStrategyRef ` and `validateRuntimeFailed ` is similar with it. One of them fails, then `validationFailed` is changed to be `true`, update Build status and stop reconciling.

Reply:
@SaschaSchwarze0 thanks, I think is a fair request, seems more clean. Let me check.

Reply:
I change this, we dont use the global var anymore, thanks for bringing this.